src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -439,6 +459,86 @@ public List<BulkLoadHistoryModel> getLoadHistory(
     }
   }
 
+  @FunctionalInterface
+  public interface IamAuthorizedCall {
+    void get() throws IamForbiddenException;
+  }
+
+  /**
+   * Throw if the user cannot access the dataset via SAM permissions.
+   *
+   * @param iamAuthorizedCall throws if dataset inaccessible via SAM permissions
+   */
+  void verifyDatasetAccessible(IamAuthorizedCall iamAuthorizedCall) {
+    try {
+      iamAuthorizedCall.get();
+    } catch (Exception iamEx) {
+      throw new ForbiddenException(
+          "Error accessing dataset: see errorDetails",
+          Collections.singletonList(iamEx.getMessage()));
+    }
+  }
+
+  /** Throw if the user cannot read the dataset. */
+  public void verifyDatasetReadable(UUID datasetId, AuthenticatedUserRequest userReq) {
+    IamAuthorizedCall canRead =
+        () ->
+            iamService.verifyAuthorization(
+                userReq, IamResourceType.DATASET, datasetId.toString(), IamAction.READ_DATA);
+    verifyDatasetAccessible(canRead);
+  }
+
+  public DatasetDataModel retrieveData(
+      AuthenticatedUserRequest userRequest,
+      UUID datasetId,
+      String tableName,
+      int limit,
+      int offset,
+      String sort,
+      SqlSortDirection direction,
+      String filter) {
+    Dataset dataset = retrieve(datasetId);
+
+    DatasetTable table =
+        dataset
+            .getTableByName(tableName)
+            .orElseThrow(
+                () ->
+                    new DatasetDataException(
+                        "No dataset table exists with the name: " + tableName));
+
+    if (!sort.equalsIgnoreCase(PDAO_ROW_ID_COLUMN)) {
```

Comment:
I actually meant to go back and remove this call altogether. This is copied from the snapshot preview code and snapshot preview currently doesn't include the row id column in its results.

Reply:
Nevermind! I still need this condition because the row id is not included in the list of columns. 
But, to your original question - since sort has a default value, we should be fine.